components/brave_mobile_subscription/renderer/android/subscription_render_frame_observer.cc

```diff
@@ -0,0 +1,201 @@
+                           weak_factory_.GetWeakPtr()));
+      }
+#endif
+      break;
+    case Product::kLeo:
+#if BUILDFLAG(ENABLE_AI_CHAT)
+      if (ai_chat_subscription_.is_bound()) {
+        ai_chat_subscription_->GetPurchaseToken(
+            base::BindOnce(&SubscriptionRenderFrameObserver::OnGetPurchaseToken,
+                           weak_factory_.GetWeakPtr()));
+      }
+#endif
+      break;
+    default:
+      NOTREACHED();
+  }
+}
+
+void SubscriptionRenderFrameObserver::OnGetPurchaseToken(
+    const std::string& purchase_token) {
+  if (!IsAllowed()) {
+    return;
+  }
+  auto* frame = render_frame();
+  if (frame) {
+    if (IsValueAllowed(purchase_token)) {
+      DCHECK(product_ != Product::kNone);
+      std::string_view receipt_var_name;
+      switch (product_) {
+        case Product::kVPN:
+          receipt_var_name = "braveVpn.receipt";
+          break;
+        case Product::kLeo:
+          receipt_var_name = "braveLeo.receipt";
+          break;
+        default:
+          NOTREACHED();
+      }
+      std::u16string set_local_storage = base::UTF8ToUTF16(
+          base::StrCat({"window.localStorage.setItem(\"", receipt_var_name,
+                        "\", \"", purchase_token, "\");"}));
+      frame->ExecuteJavaScript(set_local_storage);
+    }
+  }
+}
+
+std::string SubscriptionRenderFrameObserver::ExtractParam(
+    const GURL& url,
+    const std::string& name) const {
+  url::Component query(0, static_cast<int>(url.query_piece().length())), key,
+      value;
+  while (url::ExtractQueryKeyValue(url.query_piece().data(), &query, &key,
+                                   &value)) {
+    std::string_view key_str = url.query_piece().substr(key.begin, key.len);
+    if (key_str != name) {
+      continue;
+    }
+
+    return std::string(url.query_piece().substr(value.begin, value.len));
+  }
+  return std::string();
+}
+
+bool SubscriptionRenderFrameObserver::IsValueAllowed(
+    const std::string& purchase_token) const {
+  if (purchase_token.length() > 0) {
+    // Don't allow " in purchase token.
+    // See https://github.com/brave/brave-browser/issues/27524
+    std::size_t found = purchase_token.find("\"");
+    if (found == std::string::npos) {
+      return true;
+    }
+  }
+  return false;
+}
+
+bool SubscriptionRenderFrameObserver::IsAllowed() {
+#if BUILDFLAG(ENABLE_BRAVE_VPN)
+  DCHECK(brave_vpn::IsBraveVPNFeatureEnabled());
+#endif
+
+  if (!skus::IsSafeOrigin(render_frame()->GetWebFrame()->GetSecurityOrigin())) {
+    return false;
+  }
+
+  GURL current_url(
+      render_frame()->GetWebFrame()->GetDocument().Url().GetString().Utf8());
+
+  std::string intent = ExtractParam(current_url, kIntentParamName);
+  std::string product = ExtractParam(current_url, kProductParamName);
+  if (product == kProductVPNParamValue) {
+    product_ = Product::kVPN;
+  } else if (product == kProductLeoParamValue) {
+    product_ = Product::kLeo;
```

Comment:
so this means only one `product_` can be active at a time in this observer? if that's the case, then `EnsureConnected` should only bind to a single endpoint using the `product_` value.

Reply:
yes, it's a good point, I missed that, corrected.